test/e2e/page-objects/flows/sign.flow.ts

```diff
@@ -0,0 +1,164 @@
+): Promise<void> => {
+  const testDapp = new TestDapp(driver);
+  await testDapp.check_pageIsLoaded();
+  await testDapp.signTypedDataV3();
+  if (!isSyncFlow) {
+    await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+    await new SnapSimpleKeyringPage(driver).approveRejectSnapAccountTransaction(
+      approveTransaction,
+    );
+  }
+  if ((!isSyncFlow && approveTransaction) || isSyncFlow) {
+    await testDapp.check_successSignTypedDataV3(publicAddress);
+  } else {
+    await testDapp.check_failedSignTypedDataV3(
+      'Error: Request rejected by user or snap.',
+    );
+  }
+};
+
+/**
+ * This function initiates the steps for a signTypedDataV4 with snap account on test dapp.
+ *
+ * @param driver - The webdriver instance.
+ * @param publicAddress - The public address of the snap account.
+ * @param isSyncFlow - Indicates whether synchronous approval option is on for the snap. Defaults to true.
+ * @param approveTransaction - Indicates whether the transaction should be approved. Defaults to true.
+ */
+export const signTypedDataV4WithSnapAccount = async (
+  driver: Driver,
+  publicAddress: string,
+  isSyncFlow: boolean = true,
+  approveTransaction: boolean = true,
+): Promise<void> => {
+  const testDapp = new TestDapp(driver);
+  await testDapp.check_pageIsLoaded();
+  await testDapp.signTypedDataV4();
+  if (!isSyncFlow) {
+    await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+    await new SnapSimpleKeyringPage(driver).approveRejectSnapAccountTransaction(
+      approveTransaction,
+    );
+  }
+  if ((!isSyncFlow && approveTransaction) || isSyncFlow) {
+    await testDapp.check_successSignTypedDataV4(publicAddress);
+  } else {
+    await testDapp.check_failedSignTypedDataV4(
+      'Error: Request rejected by user or snap.',
+    );
+  }
+};
+
+/**
+ * This function initiates the steps for a signPermit with snap account on test dapp.
+ *
+ * @param driver - The webdriver instance.
+ * @param publicAddress - The public address of the snap account.
+ * @param isSyncFlow - Indicates whether synchronous approval option is on for the snap. Defaults to true.
+ * @param approveTransaction - Indicates whether the transaction should be approved. Defaults to true.
+ */
+export const signPermitWithSnapAccount = async (
+  driver: Driver,
+  publicAddress: string,
+  isSyncFlow: boolean = true,
+  approveTransaction: boolean = true,
+): Promise<void> => {
+  const testDapp = new TestDapp(driver);
+  await testDapp.check_pageIsLoaded();
+  await testDapp.signPermit();
+  if (!isSyncFlow) {
+    await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+    await new SnapSimpleKeyringPage(driver).approveRejectSnapAccountTransaction(
+      approveTransaction,
+    );
+  }
+  if ((!isSyncFlow && approveTransaction) || isSyncFlow) {
+    await testDapp.check_successSignPermit(publicAddress);
+  } else {
+    await testDapp.check_failedSignPermit(
+      'Error: Request rejected by user or snap.',
+    );
+  }
+};
```

Comment:
not blocking, just wondering if we should abstract away the shared logic on a common sign method or we prefer it like this for maybe better clarity? Both seem fine to me

Reply:
I have implemented it like this because I feel it's the easiest to use. It might be awkward to pass the sign type as a parameter, and we might have signatures that don't follow the common flow in the future, which would require adding conditional statements here and there. However, what you suggest also makes sense. I think we can keep it like this for now, and we can always improve it later if we feel that abstracting more is better.

Check how the original implementation here,I feel it's a bit complex:
https://github.com/MetaMask/metamask-extension/blob/develop/test/e2e/accounts/common.ts#L283

Reply:
@seaona @chloeYue I wrote these tests originally to be modular, abstracted, sharing logic, and DRY.  Now there's tons and tons of repeated logic 🙁

Reply:
Hi @HowardBraham, Thanks for the feedback! I understand your point. When I migrated to POM, I separated the signature functions to create ready-to-use methods for each signature type. The reason is that the original implementation asserted on the selectors and had complex logic, making it difficult to follow and understand each conditional statement. By creating these ready-to-use functions, users can call a function for a specific type of signature directly, making it straightforward to know what happens when executing a specific type of signature. DRY does not always align well with POM unfortunately, so I chose to prioritize ease of use and understanding.
However, I'm open to changing the implementation. Since all the page classes and functions have been created, we can combine all the sign functions into one and take the signature type as a parameter. The disadvantage is that this function may become more complex with future changes and implementations for signatures. But i'm open to change it.